src/pulse3D/magnet_finding.py

```diff
@@ -0,0 +1,206 @@
+        r = -np.asarray([xpos[magnet], ypos[magnet], zpos[magnet]]) + manta
+        # Kevin (12/1/21): Calculate the euclidian distance from the magnet to a given sensor
+        r_abs = np.sqrt(np.sum(r ** 2, axis=1))
+        # Kevin (12/1/21): simulate fields at sensors using dipole model for each magnet
+        for field in range(0, len(r)):
+            fields[field] += 3 * r[field] * np.dot(moment_vectors, r[field]) / r_abs[field] ** 5 - moment_vectors / r_abs[field] ** 3
+    # Kevin (12/1/21): Reshaping to match the format of the data coming off the mantarray
+    return fields.reshape((1, 3 * len(r)))[0] / DIPOLE_MODEL_FACTOR
+
+
+def objective_function_ls(
+    pos: NDArray[(1, Any), float],
+    b_meas: NDArray[(1, Any), float],
+    manta: NDArray[(72, 3), float],
+    num_active_module_ids: int
+) -> NDArray[(1, Any), float]:
+    """Cost function to be minimized by the least squares."""
+    pos = pos.reshape(NUM_PARAMS, num_active_module_ids)
+    x = pos[0]
+    y = pos[1]
+    z = pos[2]
+    theta = pos[3]
+    phi = pos[4]
+    remn = pos[5]
+
+    b_calc = meas_field(x, y, z, theta, phi, remn, manta, num_active_module_ids)
+    return b_calc - b_meas
+
+
+def get_positions(data: NDArray[(24, 3, 3, Any), float]) -> Dict[str, NDArray[(1, Any), float]]:
+    """Generate initial guess data and run least squares optimizer on instrument data to get magnet positions.
+
+    Takes an array indexed as [well, sensor, axis, timepoint]
+    Data should be the difference of the data with plate on the instrument and empty plate calibration data
+    Assumes 3 active sensors for each well, that all active wells have magnets, and that all magnets have the well beneath them active
+    """
+    # Tanner (12/3/21): hardcoding these for now. Could be made constants if appropriate
+    num_sensors = 3
+    num_axes = 3
+
+    # Tanner (12/2/21): Every well/sensor/axis will always be active as of now
+    num_active_module_ids = 24
+    active_module_ids = list(range(num_active_module_ids))
+
+    # Kevin (12/1/21): Manta gives the locations of all active sensors on all arrays with respect to a common point
+    # computing the locations of each centrally located point about which each array is to be distributed,
+    # for the purpose of offsetting the values in triad by the correct well spacing
+    # The values in "triad" and "manta" relate to layout of the board itself so they don't change at all so long as the board doesn't
+    triad = SENSOR_DISTANCES_FROM_CENTER_POINT.copy()
+    manta = np.empty((triad.shape[0] * num_active_module_ids, triad.shape[1]))
+    for module_id in range(0, num_active_module_ids):
+        module_slice = slice(module_id * triad.shape[0], (module_id + 1) * triad.shape[0])
+        manta[module_slice, :] = triad + module_id // WELLS_PER_ROW * WELL_VERTICAL_SPACING + (module_id % WELLS_PER_ROW) * WELL_HORIZONTAL_SPACING
+    print(manta.shape)
```

Comment:
print is fine for testing/debugging but we want to eventually use logger in all the pulse3d code base

Reply:
Oops, forgot to remove some of the prints